`patch-hub`: Revise 'Patchsets Details and Actions' screen

davidbtadokoro opened the pull request:
~~In the `upstream-patches-ui` feature, when showing patchset informations and actions available, there are two different functions, `show_series_details` and `show_bookmarked_series_details`, that are used depending if the patchset in question was a bookmarked one or not. The implementation of both is virtually the same and adding an action that applies for both patchset types means duplicated code.~~

~~With that in mind, this commit unifies both functions into the `show_series_details` function. The function now accepts a third argument named `show_bookmarked_patchset` that, when set, indicates that the patchset in question is a bookmarked one.~~

~~Closes: #860.~~

[EDIT]

This PR evolved to completely generalizing showing patchsets details and actions. In other words, besides refactoring (mostly) `src/ui/patch_hub/series_details.sh` for development purposes, it fixes some behavior pointed out in #879 and #880, like having to switch menus to unbookmarking a patchset and inconsistent checkmarks for patchset actions.

Commit:
src: ui: patch_hub: series_details: Unify show series functions

In the `patch-hub` feature, when showing patchset informations and
actions available, there are two different functions,
`show_series_details` and `show_bookmarked_series_details`, that are
used depending if the patchset in question was a bookmarked one or not.
The implementation of both is virtually the same and adding an action
that applies for both patchset types means duplicated code.

With that in mind, this commit unifies both functions into the
`show_series_details` function.

The `show_series_details` and `show_bookmarked_series_details` where
responsible for fetching the raw data of the patchset in question. To
make `show_series_details` more generic and clean, this responsability
is passed to the caller, as the function now accepts two arguments:
- The patchset raw data;
- A boolean indicating if the patchset is a bookmarked one or not.

Note: The boolean is only used to define which actions are available for
each case. As we plan to also unify the actions available, this argument
is sure to be removed in the future and the function shall be more
generic.

Closes: #860.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

melissawen commented:
Should this PR wait for #877 ? If so, can you put a dependency here to avoid conflicts? I plan to review `patch-hub` renaming on Wed. Another option is to develop this PR already on top of `patch-hub` changes and mention the PR dependency here anyway.

davidbtadokoro commented:
Hi @melissawen and thanks for your interest! To tell you the truth, I was planning on rebasing either this PR or the other one depending on which was closed first. But I think I'm going to go with your second approach of developing this PR on top of the other changes. I think that the commit of #877 would also be present in this PR. If so, is this a problem?

davidbtadokoro commented:
Hi @melissawen, thanks for the merge of #877 😄 . I have rebased this PR against those changes.

magalilemes commented on src/ui/patch_hub/series_details.sh:
```diff
@@ -1,102 +1,63 @@
 {
-  local patch_index="$1"
-  local -n _target_patch_metadata="$2"
+  local series_index="$1"
+  local -n _target_patch_metadata
+  local show_bookmarked_patchset="$3"
```
I don't get this part. Is the second argument of this function irrelevant? If so, wouldn't it be neater to rearrange things as to have just 2 arguments, instead of 3?

davidbtadokoro replied:
Hi @magalilemes.
Really sorry for not responding 😨.
Thanks for the heads up. Fixed it.

magalilemes commented on src/ui/patch_hub/series_details.sh:
```diff
@@ -1,102 +1,63 @@
+    action_list=('Bookmark' 'Download')
 
-  action_list=('Unbookmark')
+    _target_patch_metadata="$2"
```
found it :-)

rodrigosiqueira replied:
Assigning the second parameter here is not intuitive and breaks our code convention. Could you rethink the logic in a way that the second parameter is assigned at the beginning of this function?

Commit:
src: lib: lore: Revise patchset entity ID attribute

In the `kw patch-hub` feature, we make use of a database entity to
represent a patchset (a set of patches related to a patch series). An
instance of a patchset must have a minimal set of attributes, including
an ID (primary key) that uniquely identifies it. We use this entity for
showing details and running actions of patchsets in a mailing list and
in the local bookmarked database. However, the ID attribute is the
SHA256 value of all other attributes, which is an unnecessary and
unstable approach as a patchset URL in the lore.kernel.org public inbox
already uniquely identifies it. This approach also difficults
implementation, as we need to calculate the SHA256 taking care to use
only the minimal set attributes.

This commit removes the manufactured ID attribute and substitutes it for
the patchset URL in lore.kernel.org. This commit also substitutes some
functions and variables names with "series" or "patch" that references a
patchset instance to "patchset", as "series" is a vague term and "patch"
is imprecise as it may refer to a patchset with more than one patch
(these changes were only made where the ID substitution had an impact).

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: patch_hub_core: Remove unnecessary `previous_screen` argument of `list_patches`

The function `list_patches` uses an argument named `previous_screen` to
determine where to fetch the patchset metadata and which screen to call
next.

This commit removes this argument because `list_patches` can obtain the
same information through the `screen_sequence['SHOW_SCREEN']` value.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: series_details: Generalize showing of patchset

This commit completely generalizes the process of showing patchset's
details and actions. Previously, there was a heavy distinction between
a bookmarked and an unbookmarked patchset in `series_details.sh` which
not only made the code less clean, but also resulted in UX problems like
having to move across menus to unbookmark a patchset that was bookmarked
by accident.

The generalization is done by making invisible the type of patchset to
`series_details.sh`, and determining the actions to take (download/
remove download/bookmark/remove bookmark) by the transitions action
statuses.

Before, only actions with a set status (asterisk inside brackets) were
accounted for, which resulted in bugs and imposed limitations. For
example, having

    [*] Download
    [*] Bookmark

and hitting the OK button resulted in the actions being run again.

Now, this redundancy can be avoid. Also, because we don't distinguish
types of patchsets, the 'Unbookmark' action was removed and it is now
done by hitting the OK button after the transition

    [*] Bookmark ---> [ ] Bookmark

which possibilitates unbookmarking a patchset that was bookmarked by
accident without switching menus. The same is valid for removing the
downloaded patchset .mbx files.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: series_details: Add documentation and tests to action handlers

In the file, there is a handler function to each possible action. Those
are really important, because, if they silently fail, for example, the
user may think the patchset has been downloaded or bookmarked where in
fact it wasn't. However, there is no documentation or unit tests for
these.

With that in mind, this commit adds documentation and tests for each of
the current 4 action handlers.

Note: References to patchsets with 'series' are changed to 'patchset'
for clarity, as mentioned in previous commits.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: patchset_details_and_actions: Rename file

The filename `series_details.sh` is both imprecise and incomplete, as
'series' is vague in relation to 'patchset' and the file deals with both
details and actions of a patchset.

As such, this commit renames the file from `series_details` to
`patchset_details_and_actions`.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: ui: patch_hub: patchset_details_and_actions: Make 'Download' independent

When bookmarking a patch, the 'Download' action would also be checked
(may cause confusion to the user). Also this action doesn't really have
a clear purpose, as it does the same but less than bookmarking a patch.

In this context, this commit changes the 'Download' action to the
'Download to a specific directory' action, which allows the user to
select a directory to the download on the fly.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented on src/ui/patch_hub/series_details.sh:
```diff
@@ -1,49 +1,55 @@
-  local patch_index="$1"
-  local -n _target_patch_metadata="$2"
+  local raw_patchset="$1"
+  local show_bookmarked_patchset="$2"
```
Hi @rodrigosiqueira, thanks a lot for the feedback.
You are right, this boolean is kind of confusing. In any case, in the next commits I totally remove this parameter as this function doesn't differentiate bookmarked and non-bookmarked patchsets anymore.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -647,55 +653,55 @@ function get_bookmarked_series_by_index()
-  _series['patch_id']="${columns[7]}"
-  _series['timestamp']="${columns[8]}"
+  IFS="${SEPARATOR_CHAR}" read -ra columns <<< "${raw_patchset}"
+  _patchset['patchset_author']="${columns[0]}"
+  _patchset['author_email']="${columns[1]}"
+  _patchset['patchset_version']="${columns[2]}"
+  _patchset['total_patches']="${columns[3]}"
+  _patchset['patchset_title']="${columns[4]}"
+  _patchset['patchset_url']="${columns[5]}"
+  _patchset['download_dir_path']="${columns[6]}"
+  _patchset['timestamp']="${columns[7]}"
```
Nice idea! Going to open an issue about it.

davidbtadokoro commented:
> @davidbtadokoro could you rebase this change?

@rodrigosiqueira done 👍

rodrigosiqueira commented:
Merged! Thanks
